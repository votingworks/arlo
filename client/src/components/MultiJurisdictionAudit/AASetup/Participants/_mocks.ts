import { readFileSync } from 'fs'
import { join } from 'path'

const jurisdictionFile = new File(
  [
    readFileSync(
      join(
        __dirname,
        '../../../../../public/sample_jurisdiction_filesheet.csv'
      ),
      'utf8'
    ),
  ],
  'jurisdictions.csv',
  { type: 'text/csv' }
)

export default jurisdictionFile
