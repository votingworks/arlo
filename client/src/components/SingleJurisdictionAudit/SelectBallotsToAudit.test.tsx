import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'
import { toast } from 'react-toastify'
import SelectBallotsToAudit from './SelectBallotsToAudit'
import { statusStates, ballotManifest } from './_mocks'
import { regexpEscape } from '../testUtilities'
import * as utilities from '../utilities'

const apiMock: jest.SpyInstance<
  ReturnType<typeof utilities.api>,
  Parameters<typeof utilities.api>
> = jest.spyOn(utilities, 'api').mockImplementation()

const checkAndToastMock: jest.SpyInstance<
  ReturnType<typeof utilities.checkAndToast>,
  Parameters<typeof utilities.checkAndToast>
> = jest.spyOn(utilities, 'checkAndToast').mockReturnValue(false)
const toastSpy = jest.spyOn(toast, 'error').mockImplementation()

async function inputAndSubmitForm() {
  const getStatusMock = jest
    .fn()
    .mockImplementationOnce(async () => statusStates.ballotManifestProcessed) // the POST to /election/{electionId}/audit/status after jurisdictions
    .mockImplementation(async () => statusStates.completeInFirstRound) // the POST to /election/{electionId}/audit/status after manifest
  const updateAuditMock = jest
    .fn()
    .mockImplementationOnce(async () => statusStates.completeInFirstRound) // the POST to /election/{electionId}/audit/status after manifest

  const { getByLabelText, getByText } = render(
    <SelectBallotsToAudit
      audit={statusStates.sampleSizeOptions}
      isLoading={false}
      setIsLoading={jest.fn()}
      updateAudit={updateAuditMock}
      getStatus={getStatusMock}
      electionId="1"
    />
  )

  const manifestInput = getByLabelText('Select manifest...')
  fireEvent.change(manifestInput, { target: { files: [] } })
  fireEvent.blur(manifestInput)
  await waitFor(() => {
    expect(getByText('You must upload a manifest')).toBeTruthy()
  })
  fireEvent.change(manifestInput, { target: { files: [ballotManifest] } })

  const auditBoardInput: HTMLElement = getByLabelText(
    new RegExp(regexpEscape('Set the number of audit boards you wish to use.')),
    { selector: 'select' }
  )
  fireEvent.change(auditBoardInput, { target: { selectedIndex: 1 } })

  // const boardOneNameInput: HTMLElement = getByTestId('audit-name-0')
  // fireEvent.change(boardOneNameInput, { target: { value: 'Board One' } }) // removed until custom audit board name feature is added again

  const sampleSizeInput = getByLabelText(
    '379 samples (80% chance of reaching risk limit and completing the audit in one round)'
  )
  fireEvent.click(sampleSizeInput, { bubbles: true })

  const submitButton = getByText('Select Ballots To Audit')
  fireEvent.click(submitButton, { bubbles: true })

  return [getStatusMock, updateAuditMock]
}

beforeEach(() => {
  apiMock.mockClear()
  toastSpy.mockClear()
  checkAndToastMock.mockClear()
})

describe('SelectBallotsToAudit', () => {
  it('renders correctly', () => {
    const { container, rerender } = render(
      <SelectBallotsToAudit
        audit={statusStates.contestFirstRound}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        getStatus={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()

    rerender(
      <SelectBallotsToAudit
        audit={statusStates.contestFirstRound}
        isLoading
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        getStatus={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('has radio for selecting sampleSize', () => {
    const { getByText, getByLabelText } = render(
      <SelectBallotsToAudit
        audit={statusStates.sampleSizeOptions}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        getStatus={jest.fn()}
        electionId="1"
      />
    )

    // all options should be present
    expect(getByText('BRAVO Average Sample Number: 269 samples')).toBeTruthy()
    expect(
      getByText(
        '379 samples (80% chance of reaching risk limit and completing the audit in one round)'
      )
    ).toBeTruthy()
    expect(getByText('78 samples')).toBeTruthy()
    expect(
      getByText('Enter your own sample size (not recommended)')
    ).toBeTruthy()

    // correct default should be selected
    expect(
      getByLabelText('BRAVO Average Sample Number: 269 samples')
    ).toBeChecked()
  })

  it('conditionally shows custom text input and submits', async () => {
    const getStatusMock = jest
      .fn()
      .mockImplementationOnce(async () => statusStates.jurisdictionsInitial) // the POST to /election/{electionId}/audit/status after jurisdictions
      .mockImplementation(async () => statusStates.ballotManifestProcessed) // the POST to /election/{electionId}/audit/status after manifest
    const updateAuditMock = jest
      .fn()
      .mockImplementationOnce(async () => statusStates.ballotManifestProcessed) // the POST to /election/{electionId}/audit/status after manifest
    apiMock.mockImplementation(async () => {})

    const {
      getByText,
      queryAllByTestId,
      getByTestId,
      getByLabelText,
      queryAllByText,
    } = render(
      <SelectBallotsToAudit
        audit={statusStates.sampleSizeOptions}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={updateAuditMock}
        getStatus={getStatusMock}
        electionId="1"
      />
    )

    expect(queryAllByTestId('customSampleSize[contest-1]').length).toBe(0)
    const customRadio = getByText(
      'Enter your own sample size (not recommended)'
    )
    fireEvent.click(customRadio, { bubbles: true })
    await waitFor(() => {
      expect(queryAllByTestId('customSampleSize[contest-1]').length).toBe(1)
    })
    fireEvent.click(getByText('BRAVO Average Sample Number: 269 samples'), {
      bubbles: true,
    })
    await waitFor(() => {
      expect(queryAllByTestId('customSampleSize[contest-1]').length).toBe(0)
    })
    fireEvent.click(customRadio, { bubbles: true })
    await waitFor(() => {
      expect(queryAllByTestId('customSampleSize[contest-1]').length).toBe(1)
    })

    const customInput = getByTestId('customSampleSize[contest-1]')
    fireEvent.change(customInput, { target: { value: '3000' } })
    fireEvent.blur(customInput)
    await waitFor(() => {
      expect(
        getByText('Must be less than or equal to the total number of ballots')
      ).toBeTruthy()
    })
    fireEvent.change(customInput, { target: { value: '11' } })
    fireEvent.blur(customInput)
    await waitFor(() => {
      expect(
        queryAllByText(
          'Must be less than or equal to the total number of ballots'
        ).length
      ).toBe(0)
    })

    const manifestInput = getByLabelText('Select manifest...')
    fireEvent.change(manifestInput, { target: { files: [ballotManifest] } })

    const submitButton = getByText('Select Ballots To Audit')
    fireEvent.click(submitButton, { bubbles: true })

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(3)

      expect(apiMock.mock.calls[0][0]).toMatch(
        /\/election\/[^/]+\/audit\/sample-size/
      )
      expect(apiMock.mock.calls[0][1]).toMatchObject({
        method: 'POST',
        body: JSON.stringify({
          size: '11',
        }),
        headers: {
          'Content-Type': 'application/json',
        },
      })

      expect(getStatusMock).toBeCalledTimes(2)
      expect(updateAuditMock).toBeCalledTimes(1)
      expect(checkAndToastMock).toBeCalledTimes(3)
    })
  })

  it('bails if the manifest failed to start processing', async () => {
    const getStatusMock = jest
      .fn()
      .mockImplementation(async () => statusStates.jurisdictionsInitial) // the POST to /election/{electionId}/audit/status after jurisdictions
    const updateAuditMock = jest
      .fn()
      .mockImplementationOnce(async () => statusStates.ballotManifestProcessed) // the POST to /election/{electionId}/audit/status after manifest
    apiMock.mockImplementation(async () => {})

    const { getByText, getByLabelText } = render(
      <SelectBallotsToAudit
        audit={statusStates.sampleSizeOptions}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={updateAuditMock}
        getStatus={getStatusMock}
        electionId="1"
      />
    )

    const manifestInput = getByLabelText('Select manifest...')
    fireEvent.change(manifestInput, { target: { files: [ballotManifest] } })

    const submitButton = getByText('Select Ballots To Audit')
    fireEvent.click(submitButton, { bubbles: true })

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(3)

      expect(getStatusMock).toBeCalledTimes(2)
      expect(updateAuditMock).toBeCalledTimes(0)
      expect(checkAndToastMock).toBeCalledTimes(3)
    })
  })

  it('does not bail if the manifest processing errors', async () => {
    const getStatusMock = jest
      .fn()
      .mockImplementation(async () => statusStates.ballotManifestProcessError) // the POST to /election/{electionId}/audit/status after jurisdictions
    const updateAuditMock = jest
      .fn()
      .mockImplementationOnce(async () => statusStates.ballotManifestProcessed) // the POST to /election/{electionId}/audit/status after manifest
    apiMock.mockImplementation(async () => {})

    const { getByText, getByLabelText } = render(
      <SelectBallotsToAudit
        audit={statusStates.sampleSizeOptions}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={updateAuditMock}
        getStatus={getStatusMock}
        electionId="1"
      />
    )

    const manifestInput = getByLabelText('Select manifest...')
    fireEvent.change(manifestInput, { target: { files: [ballotManifest] } })

    const submitButton = getByText('Select Ballots To Audit')
    fireEvent.click(submitButton, { bubbles: true })

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(3)

      expect(getStatusMock).toBeCalledTimes(2)
      expect(updateAuditMock).toBeCalledTimes(1)
      expect(checkAndToastMock).toBeCalledTimes(3)
    })
  })

  it('handles server timeout on ballot manifest processing', async () => {
    const dateIncrementor = (function* incr() {
      let i = 10
      while (true) {
        i += 130000
        yield i
      }
    })()
    const dateSpy = jest
      .spyOn(Date, 'now')
      .mockImplementation(() => dateIncrementor.next().value)

    const getStatusMock = jest
      .fn()
      .mockImplementation(async () => statusStates.jurisdictionsInitial) // the POST to /election/{electionId}/audit/status after jurisdictions
    const updateAuditMock = jest
      .fn()
      .mockImplementationOnce(async () => statusStates.ballotManifestProcessed) // the POST to /election/{electionId}/audit/status after manifest
    apiMock.mockImplementation(async () => {})

    const { getByText, getByLabelText } = render(
      <SelectBallotsToAudit
        audit={statusStates.sampleSizeOptions}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={updateAuditMock}
        getStatus={getStatusMock}
        electionId="1"
      />
    )

    const manifestInput = getByLabelText('Select manifest...')
    fireEvent.change(manifestInput, { target: { files: [ballotManifest] } })

    const submitButton = getByText('Select Ballots To Audit')
    fireEvent.click(submitButton, { bubbles: true })

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(3)
      expect(dateSpy).toBeCalled()
      expect(updateAuditMock).toBeCalledTimes(0)
      expect(toastSpy).toBeCalledTimes(1)
      expect(getStatusMock).toBeCalledTimes(2)
      expect(checkAndToastMock).toBeCalledTimes(3)
    })
  })

  it('changes sampleSize based on audit.rounds.contests.sampleSize', () => {
    const { getByLabelText } = render(
      <Router>
        <SelectBallotsToAudit
          audit={statusStates.completeInFirstRound}
          isLoading={false}
          setIsLoading={jest.fn()}
          updateAudit={jest.fn()}
          getStatus={jest.fn()}
          electionId="1"
        />
      </Router>
    )

    expect(
      getByLabelText(
        '379 samples (80% chance of reaching risk limit and completing the audit in one round)'
      )
    ).toBeChecked()
  })

  it('changes number of audits', () => {
    const { getByLabelText, container } = render(
      <SelectBallotsToAudit
        audit={statusStates.contestFirstRound}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        getStatus={jest.fn()}
        electionId="1"
      />
    )

    const auditBoardInput = getByLabelText(
      new RegExp(
        regexpEscape('Set the number of audit boards you wish to use.')
      ),
      { selector: 'select' }
    )
    expect(auditBoardInput).toBeInstanceOf(HTMLSelectElement)
    if (auditBoardInput instanceof HTMLSelectElement) {
      fireEvent.change(auditBoardInput, { target: { selectedIndex: 2 } })
      expect(auditBoardInput.selectedOptions[0].innerHTML).toBe('3')
      expect(container).toMatchSnapshot()
      fireEvent.change(auditBoardInput, { target: { selectedIndex: 1 } })
      expect(container).toMatchSnapshot()
    }
  })

  it('submits sample size, ballot manifest, and audits', async () => {
    apiMock.mockImplementation(async () => {})

    const [getStatusMock, updateAuditMock] = await inputAndSubmitForm()

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(3)

      expect(apiMock.mock.calls[0]).toMatchObject([
        expect.stringMatching(/\/election\/[^/]+\/audit\/sample-size/),
        {
          method: 'POST',
          body: JSON.stringify({
            size: '379',
          }),
          headers: {
            'Content-Type': 'application/json',
          },
        },
      ])

      {
        const [url, options] = apiMock.mock.calls[1]
        const body = options && JSON.parse(options.body as string)

        expect(url).toMatch(/\/election\/[^/]+\/audit\/jurisdictions/)
        expect(body).toMatchObject({
          // TODO fix type
          jurisdictions: [
            {
              id: expect.stringMatching(/^[-0-9a-z]+$/),
              name: 'Jurisdiction 1',
              contests: ['contest-1'],
              auditBoards: [
                {
                  id: expect.stringMatching(/^[-0-9a-z]+$/),
                  name: 'Audit Board #1', // change to 'Board One' if custom audit board feature is added back again
                  members: [],
                },
                {
                  id: expect.stringMatching(/^[-0-9a-z]+$/),
                  name: 'Audit Board #2',
                  members: [],
                },
              ],
            },
          ],
        })
      }

      expect(apiMock.mock.calls[2][0]).toMatch(
        /\/election\/[^/]+\/jurisdiction\/jurisdiction-1\/manifest/
      )

      expect(getStatusMock).toBeCalledTimes(2)
      expect(updateAuditMock).toBeCalledTimes(1)
    })
  })

  it('handles api error on /audit/sample-size', async () => {
    apiMock
      .mockRejectedValueOnce({ message: 'error' })
      .mockImplementation(async () => ({}))

    const [getStatusMock, updateAuditMock] = await inputAndSubmitForm()

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(1)
      expect(toastSpy).toBeCalledTimes(1)

      expect(getStatusMock).toBeCalledTimes(0)
      expect(updateAuditMock).toBeCalledTimes(0)
    })
  })

  it('handles api error on /audit/jurisdictions', async () => {
    apiMock
      .mockResolvedValueOnce(undefined)
      .mockRejectedValueOnce({ message: 'error' })
      .mockResolvedValue(undefined)

    const [getStatusMock, updateAuditMock] = await inputAndSubmitForm()

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(2)
      expect(toastSpy).toBeCalledTimes(1)

      expect(getStatusMock).toBeCalledTimes(0)
      expect(updateAuditMock).toBeCalledTimes(0)
    })
  })

  it('handles api error on /audit/jurisdiction/:id/manifest', async () => {
    apiMock
      .mockResolvedValueOnce(undefined)
      .mockResolvedValueOnce(undefined)
      .mockRejectedValueOnce({ message: 'error' })

    const [getStatusMock, updateAuditMock] = await inputAndSubmitForm()

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(3)
      expect(toastSpy).toBeCalledTimes(1)

      expect(getStatusMock).toBeCalledTimes(1)
      expect(updateAuditMock).toBeCalledTimes(0)
    })
  })

  it('handles server error on /audit/sample-size', async () => {
    apiMock.mockResolvedValue(undefined)
    checkAndToastMock.mockReturnValueOnce(true).mockReturnValue(false)

    const [getStatusMock, updateAuditMock] = await inputAndSubmitForm()

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(1)
      expect(toastSpy).toBeCalledTimes(0)
      expect(checkAndToastMock).toBeCalledTimes(1)
      expect(getStatusMock).toBeCalledTimes(0)
      expect(updateAuditMock).toBeCalledTimes(0)
    })
  })

  it('handles server error on /audit/jurisdictions', async () => {
    apiMock.mockResolvedValue(undefined)
    checkAndToastMock
      .mockReturnValueOnce(false)
      .mockReturnValueOnce(true)
      .mockReturnValue(false)

    const [getStatusMock, updateAuditMock] = await inputAndSubmitForm()

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(2)
      expect(toastSpy).toBeCalledTimes(0)
      expect(checkAndToastMock).toBeCalledTimes(2)
      expect(getStatusMock).toBeCalledTimes(0)
      expect(updateAuditMock).toBeCalledTimes(0)
    })
  })

  it('handles server error on /audit/jurisdiction/:id/manifest', async () => {
    apiMock.mockResolvedValue(undefined)
    checkAndToastMock
      .mockReturnValueOnce(false)
      .mockReturnValueOnce(false)
      .mockReturnValueOnce(true)
      .mockReturnValue(false)

    const [getStatusMock, updateAuditMock] = await inputAndSubmitForm()

    await waitFor(() => {
      expect(apiMock).toBeCalledTimes(3)
      expect(toastSpy).toBeCalledTimes(0)
      expect(checkAndToastMock).toBeCalledTimes(3)
      expect(getStatusMock).toBeCalledTimes(1)
      expect(updateAuditMock).toBeCalledTimes(0)
    })
  })

  it('uses the highest prob value from duplicate sampleSizes', () => {
    statusStates.contestFirstRound.rounds[0].contests[0].sampleSizeOptions = [
      { size: 30, prob: 0.8, type: null },
      { size: 30, prob: 0.9, type: null },
    ]
    const { queryAllByText } = render(
      <SelectBallotsToAudit
        audit={statusStates.contestFirstRound}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        getStatus={jest.fn()}
        electionId="1"
      />
    )

    expect(
      queryAllByText(
        '30 samples (90% chance of reaching risk limit and completing the audit in one round)'
      ).length
    ).toBe(1)
  })

  it('does not display duplicate sampleSize options', () => {
    const statusState = { ...statusStates.contestFirstRound }
    statusState.rounds[0].contests[0].sampleSizeOptions = [
      { size: 30, prob: null, type: null },
      { size: 30, prob: null, type: null },
    ]
    const { queryAllByText } = render(
      <SelectBallotsToAudit
        audit={statusState}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        getStatus={jest.fn()}
        electionId="1"
      />
    )

    expect(queryAllByText('30 samples').length).toBe(1)
  })
})
