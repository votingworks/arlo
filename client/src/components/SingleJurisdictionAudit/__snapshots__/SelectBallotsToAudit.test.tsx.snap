// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBallotsToAudit changes number of audits 1`] = `
<div>
  <form
    data-testid="form-two"
    id="fillFormTwo"
  >
    <hr />
    <div
      class="sc-ifAKCX dOqzfM"
    >
      <h2
        class="bp3-heading sc-bxivhb jFEWwS"
      />
      <h2
        class="bp3-heading sc-bxivhb jFEWwS"
      >
        Select Ballots to Audit
      </h2>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <h3
          class="bp3-heading sc-htpNat cNFUHJ"
        >
          Audit Boards
        </h3>
        <label
          for="auditBoards"
        >
          Set the number of audit boards you wish to use.
          <div
            class="bp3-html-select sc-fjdhpX jqDqiz"
          >
            <select
              field="[object Object]"
              form="[object Object]"
              id="auditBoards"
            >
              <option>
                1
              </option>
              <option>
                2
              </option>
              <option>
                3
              </option>
              <option>
                4
              </option>
              <option>
                5
              </option>
              <option>
                6
              </option>
              <option>
                7
              </option>
              <option>
                8
              </option>
              <option>
                9
              </option>
              <option>
                10
              </option>
              <option>
                11
              </option>
              <option>
                12
              </option>
              <option>
                13
              </option>
              <option>
                14
              </option>
              <option>
                15
              </option>
            </select>
            <span
              class="bp3-icon bp3-icon-double-caret-vertical"
              icon="double-caret-vertical"
            >
              <svg
                data-icon="double-caret-vertical"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  double-caret-vertical
                </desc>
                <path
                  d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </div>
        </label>
        <div
          class="sc-bwzfXH izcbHX"
        >
          Audit boards will enter data about each ballot:
        </div>
        <div
          class="sc-jzJRlG Jpdxv"
        >
          <div
            class="sc-cSHVUG dyClCA"
          />
          <div
            class="sc-cSHVUG dyClCA"
          />
          <div
            class="sc-cSHVUG dyClCA"
          />
        </div>
      </div>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <h3
          class="bp3-heading sc-htpNat cNFUHJ"
        >
          Ballot Manifest
        </h3>
        <div
          class="sc-bwzfXH izcbHX"
        >
          Click "Browse" to choose the appropriate Ballot Manifest file from your computer
        </div>
        <label
          class="bp3-file-input"
        >
          <input
            accept=".csv"
            name="manifest"
            type="file"
          />
          <span
            class="bp3-file-upload-input"
          >
            Select manifest...
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-gzVnrw ecjOPR"
    >
      <button
        class="bp3-button bp3-intent-primary sc-bZQynM iFVUEu"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Select Ballots To Audit
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`SelectBallotsToAudit changes number of audits 2`] = `
<div>
  <form
    data-testid="form-two"
    id="fillFormTwo"
  >
    <hr />
    <div
      class="sc-ifAKCX dOqzfM"
    >
      <h2
        class="bp3-heading sc-bxivhb jFEWwS"
      />
      <h2
        class="bp3-heading sc-bxivhb jFEWwS"
      >
        Select Ballots to Audit
      </h2>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <h3
          class="bp3-heading sc-htpNat cNFUHJ"
        >
          Audit Boards
        </h3>
        <label
          for="auditBoards"
        >
          Set the number of audit boards you wish to use.
          <div
            class="bp3-html-select sc-fjdhpX jqDqiz"
          >
            <select
              field="[object Object]"
              form="[object Object]"
              id="auditBoards"
            >
              <option>
                1
              </option>
              <option>
                2
              </option>
              <option>
                3
              </option>
              <option>
                4
              </option>
              <option>
                5
              </option>
              <option>
                6
              </option>
              <option>
                7
              </option>
              <option>
                8
              </option>
              <option>
                9
              </option>
              <option>
                10
              </option>
              <option>
                11
              </option>
              <option>
                12
              </option>
              <option>
                13
              </option>
              <option>
                14
              </option>
              <option>
                15
              </option>
            </select>
            <span
              class="bp3-icon bp3-icon-double-caret-vertical"
              icon="double-caret-vertical"
            >
              <svg
                data-icon="double-caret-vertical"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  double-caret-vertical
                </desc>
                <path
                  d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </div>
        </label>
        <div
          class="sc-bwzfXH izcbHX"
        >
          Audit boards will enter data about each ballot:
        </div>
        <div
          class="sc-jzJRlG Jpdxv"
        >
          <div
            class="sc-cSHVUG dyClCA"
          />
          <div
            class="sc-cSHVUG dyClCA"
          />
        </div>
      </div>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <h3
          class="bp3-heading sc-htpNat cNFUHJ"
        >
          Ballot Manifest
        </h3>
        <div
          class="sc-bwzfXH izcbHX"
        >
          Click "Browse" to choose the appropriate Ballot Manifest file from your computer
        </div>
        <label
          class="bp3-file-input"
        >
          <input
            accept=".csv"
            name="manifest"
            type="file"
          />
          <span
            class="bp3-file-upload-input"
          >
            Select manifest...
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-gzVnrw ecjOPR"
    >
      <button
        class="bp3-button bp3-intent-primary sc-bZQynM iFVUEu"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Select Ballots To Audit
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`SelectBallotsToAudit renders correctly 1`] = `
<div>
  <form
    data-testid="form-two"
    id="fillFormTwo"
  >
    <hr />
    <div
      class="sc-ifAKCX dOqzfM"
    >
      <h2
        class="bp3-heading sc-bxivhb jFEWwS"
      />
      <h2
        class="bp3-heading sc-bxivhb jFEWwS"
      >
        Select Ballots to Audit
      </h2>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <h3
          class="bp3-heading sc-htpNat cNFUHJ"
        >
          Audit Boards
        </h3>
        <label
          for="auditBoards"
        >
          Set the number of audit boards you wish to use.
          <div
            class="bp3-html-select sc-fjdhpX jqDqiz"
          >
            <select
              field="[object Object]"
              form="[object Object]"
              id="auditBoards"
            >
              <option>
                1
              </option>
              <option>
                2
              </option>
              <option>
                3
              </option>
              <option>
                4
              </option>
              <option>
                5
              </option>
              <option>
                6
              </option>
              <option>
                7
              </option>
              <option>
                8
              </option>
              <option>
                9
              </option>
              <option>
                10
              </option>
              <option>
                11
              </option>
              <option>
                12
              </option>
              <option>
                13
              </option>
              <option>
                14
              </option>
              <option>
                15
              </option>
            </select>
            <span
              class="bp3-icon bp3-icon-double-caret-vertical"
              icon="double-caret-vertical"
            >
              <svg
                data-icon="double-caret-vertical"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  double-caret-vertical
                </desc>
                <path
                  d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </div>
        </label>
        <div
          class="sc-bwzfXH izcbHX"
        >
          Audit boards will enter data about each ballot:
        </div>
        <div
          class="sc-jzJRlG Jpdxv"
        >
          <div
            class="sc-cSHVUG dyClCA"
          />
        </div>
      </div>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <h3
          class="bp3-heading sc-htpNat cNFUHJ"
        >
          Ballot Manifest
        </h3>
        <div
          class="sc-bwzfXH izcbHX"
        >
          Click "Browse" to choose the appropriate Ballot Manifest file from your computer
        </div>
        <label
          class="bp3-file-input"
        >
          <input
            accept=".csv"
            name="manifest"
            type="file"
          />
          <span
            class="bp3-file-upload-input"
          >
            Select manifest...
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-gzVnrw ecjOPR"
    >
      <button
        class="bp3-button bp3-intent-primary sc-bZQynM iFVUEu"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Select Ballots To Audit
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`SelectBallotsToAudit renders correctly 2`] = `
<div>
  <form
    data-testid="form-two"
    id="fillFormTwo"
  >
    <hr />
    <div
      class="sc-ifAKCX dOqzfM"
    >
      <h2
        class="bp3-heading sc-bxivhb jFEWwS"
      />
      <h2
        class="bp3-heading sc-bxivhb jFEWwS"
      >
        Select Ballots to Audit
      </h2>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <h3
          class="bp3-heading sc-htpNat cNFUHJ"
        >
          Audit Boards
        </h3>
        <label
          for="auditBoards"
        >
          Set the number of audit boards you wish to use.
          <div
            class="bp3-html-select sc-fjdhpX jqDqiz"
          >
            <select
              field="[object Object]"
              form="[object Object]"
              id="auditBoards"
            >
              <option>
                1
              </option>
              <option>
                2
              </option>
              <option>
                3
              </option>
              <option>
                4
              </option>
              <option>
                5
              </option>
              <option>
                6
              </option>
              <option>
                7
              </option>
              <option>
                8
              </option>
              <option>
                9
              </option>
              <option>
                10
              </option>
              <option>
                11
              </option>
              <option>
                12
              </option>
              <option>
                13
              </option>
              <option>
                14
              </option>
              <option>
                15
              </option>
            </select>
            <span
              class="bp3-icon bp3-icon-double-caret-vertical"
              icon="double-caret-vertical"
            >
              <svg
                data-icon="double-caret-vertical"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  double-caret-vertical
                </desc>
                <path
                  d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </div>
        </label>
        <div
          class="sc-bwzfXH izcbHX"
        >
          Audit boards will enter data about each ballot:
        </div>
        <div
          class="sc-jzJRlG Jpdxv"
        >
          <div
            class="sc-cSHVUG dyClCA"
          />
        </div>
      </div>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <h3
          class="bp3-heading sc-htpNat cNFUHJ"
        >
          Ballot Manifest
        </h3>
        <div
          class="sc-bwzfXH izcbHX"
        >
          Click "Browse" to choose the appropriate Ballot Manifest file from your computer
        </div>
        <label
          class="bp3-file-input"
        >
          <input
            accept=".csv"
            name="manifest"
            type="file"
          />
          <span
            class="bp3-file-upload-input"
          >
            Select manifest...
          </span>
        </label>
      </div>
    </div>
    <div
      class="bp3-spinner"
    >
      <div
        class="bp3-spinner-animation"
      >
        <svg
          height="50"
          stroke-width="8.00"
          viewBox="1.00 1.00 98.00 98.00"
          width="50"
        >
          <path
            class="bp3-spinner-track"
            d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
          />
          <path
            class="bp3-spinner-head"
            d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
            pathLength="280"
            stroke-dasharray="280 280"
            stroke-dashoffset="210"
          />
        </svg>
      </div>
    </div>
  </form>
</div>
`;
